fdbserver/CommitProxyServer.actor.cpp

```diff
@@ -3349,6 +3350,67 @@ ACTOR Future<Void> processTransactionStateRequestPart(TransactionStateResolveCon
 
 } // anonymous namespace
 
+//
+// Metrics related to the commit proxy are logged on a five second interval in
+// the `ProxyMetrics` trace.  However, it can be hard to determine workload
+// burstiness when looking at such a large time range.  This function adds much
+// more frequent logging for certain metrics to provide fine-grained insight
+// into workload patterns. The metrics logged by this function break down into
+// two categories:
+//
+//   * existing counters reported by `ProxyMetrics`
+//   * new counters that are only reported by this function
+//
+// Neither is implemented optimally, but the data collected should be helpful
+// in identifying workload patterns on the server.
+//
+// Metrics reporting by this function can be disabled by setting the
+// `BURSTINESS_METRICS_LOG_INTERVAL` knob to a negative value.
+//
+ACTOR Future<Void> logDetailedMetrics(ProxyCommitData* commitData) {
+	state int64_t commitBatchInBaseline = 0;
+	state int64_t txnCommitInBaseline = 0;
+	state int64_t mutationsBaseline = 0;
+	state int64_t mutationBytesBaseline = 0;
+	state double last_interval = now();
+
+	if (SERVER_KNOBS->BURSTINESS_METRICS_LOG_INTERVAL < 0) {
+		return Void();
+	}
+
+	loop {
+		wait(delay(SERVER_KNOBS->BURSTINESS_METRICS_LOG_INTERVAL));
+
+		int64_t commitBatchInReal = commitData->stats.commitBatchIn.getValue();
+		int64_t txnCommitInReal = commitData->stats.txnCommitIn.getValue();
+		int64_t mutationsReal = commitData->stats.mutations.getValue();
+		int64_t mutationBytesReal = commitData->stats.mutationBytes.getValue();
+
+		TraceEvent("ProxyDetailedMetrics")
+		    .detail("Elapsed", now() - last_interval)
+		    .detail("CommitBatchIn", commitBatchInReal - commitBatchInBaseline)
+		    .detail("TxnCommitIn", txnCommitInReal - txnCommitInBaseline)
+		    .detail("Mutations", mutationsReal - mutationsBaseline)
+		    .detail("MutationBytes", mutationBytesReal - mutationBytesBaseline)
+		    .detail("UniqueClients", commitData->stats.getSizeAndResetUniqueClients());
+
+		if (commitBatchInReal < commitBatchInBaseline || txnCommitInReal < txnCommitInBaseline ||
```

Comment:
It took me a while to figure out why this was here. It seems to me that if the counters have been reset by the main metrics loop, then we're better off throwing them away for that round than losing the values that were truncated.

I think a simpler way to accomplish this is
1) Record the baselines before the wait.
2) Wait.
3) Record the new values.
4) If the new values are smaller than the baseline, don't use them.

Reply:
I slightly modified 4 to also check if all of the new values are equal to the baseline, in which case nothing will be logged. I'm not 100% certain this is desirable behavior, but I think it is, as it should hopefully cut down on logging 0 values, and the zero values should be able to be inferred when looking at the data.